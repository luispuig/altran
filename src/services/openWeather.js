// @flow
const OPEN_WEATHER_MAP_URL = "https://api.openweathermap.org/data/2.5/weather?appid=99cf78b8e2240a69ab073fd37c98cea7&units=metric";

type _ApiData = {
  code: number,
  message?: string,
  main?: { temp: number }
}


const api = {
  getTempByCityId: function (id:number) {
    const encodedLocation = encodeURIComponent(String(id));
    const requestUrl = `${OPEN_WEATHER_MAP_URL}&id=${encodedLocation}`;


    return fetch(requestUrl)
        .then((response) => {
          return response.json()
        })
        .then((data:_ApiData) => {
          if (data.cod && data.message)
          {
            let error:{ message:string, origin?:string } = new Error(data.message);
            error.origin = "api";
            throw error;
          }
          else if(data.main !== undefined)
            return data.main.temp;
          else
            throw new Error('Unable to fetch weather for that location'); // throw generalistic error
        })
        .catch( (e:{ message: string, origin?: string}) =>  {
          const {origin} = e;
          if(origin === "api")
            throw e; // throw error message returned by api service
          else
            throw new Error('Unable to fetch weather for that location'); // throw generalistic error
        });

  }

};

export default api;
