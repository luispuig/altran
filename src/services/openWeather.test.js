import openWeather from "./openWeather";

it('Loads data', async () => {

  const response = { main: { temp: 25 } };
  window.fetch = jest.fn().mockImplementation( () => Promise.resolve( { json: () => response } ) );

  const result = await openWeather.getTempByCityId(123);

  expect(result).toMatchSnapshot();

});

it('Loads data fail', async () => {

  const response = { cod: 404, message: "error message" };
  window.fetch = jest.fn().mockImplementation( () => Promise.resolve( { json: () => response } ) );

  let result;
  try {
    result = await openWeather.getTempByCityId(123);

  } catch (error) {
    result = error.message;
  }

  expect(result).toMatchSnapshot();

});

it('Loads data error', async () => {

  const response = { };
  window.fetch = jest.fn().mockImplementation( () => Promise.resolve( { json: () => response } ) );

  let result;
  try {
    result = await openWeather.getTempByCityId(123);

  } catch (error) {
    result = error.message;
  }
  expect(result).toMatchSnapshot();

});
