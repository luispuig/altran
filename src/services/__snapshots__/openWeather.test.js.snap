// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loads data 1`] = `25`;

exports[`Loads data error 1`] = `"Unable to fetch weather for that location"`;

exports[`Loads data fail 1`] = `"error message"`;
